org.bndtools.launch/src/bndtools/launch/AbstractOSGiLaunchDelegate.java

```diff
@@ -171,70 +182,153 @@ public boolean buildForLaunch(ILaunchConfiguration configuration, String mode, I
+	private boolean checkExistingLaunches(ILaunchConfiguration configuration) throws CoreException {
+		IResource launchResource = LaunchUtils.getTargetResource(configuration);
+		if (launchResource == null)
+			throw new CoreException(new Status(IStatus.ERROR, PLUGIN_ID, 0,
+				"Bnd launch target was not specified or does not exist.", null));
+
+		if (!hasRunningProcess(configuration))
+			return true;
+
+		Status status = new Status(IStatus.WARNING, PLUGIN_ID, 0,
+			"One or more OSGi Frameworks have already been launched for this configuration. Additional framework instances may interfere with each other due to the shared storage directory.",
+			null);
+
+		IStatusHandler prompter = DebugPlugin.getDefault()
+			.getStatusHandler(status);
+
+		if (prompter != null) {
+			return (boolean) prompter.handleStatus(status, launchResource);
+		}
+
+		return true;
+	}
+
+	private boolean hasRunningProcess(ILaunchConfiguration configuration) throws CoreException {
+
+		IResource launchResource = LaunchUtils.getTargetResource(configuration);
+		if (launchResource == null)
+			throw new CoreException(new Status(IStatus.ERROR, PLUGIN_ID, 0,
+				"Bnd launch target was not specified or does not exist.", null));
+
+		for (ILaunch l : DebugPlugin.getDefault()
+			.getLaunchManager()
+			.getLaunches()) {
+
+			if (configuration.equals(l.getLaunchConfiguration())) {
+				if (launchResource.equals(LaunchUtils.getTargetResource(l.getLaunchConfiguration()))) {
 					for (IProcess process : l.getProcesses()) {
-						if (!process.isTerminated())
-							processCount++;
+						if (!process.isTerminated()) {
+							return true;
+						}
 					}
 				}
 			}
+		}
+		return false;
+	}
 
-			// Warn if existing processes running
-			if (processCount > 0) {
-				Status status = new Status(IStatus.WARNING, PLUGIN_ID, 0,
-					"One or more OSGi Frameworks have already been launched for this configuration. Additional framework instances may interfere with each other due to the shared storage directory.",
-					null);
-				IStatusHandler prompter = DebugPlugin.getDefault()
-					.getStatusHandler(status);
-				if (prompter != null) {
-					boolean okay = (Boolean) prompter.handleStatus(status, launchResource);
-					if (!okay)
-						return okay;
-				}
-			}
+	private boolean checkProjectErrors(ILaunchConfiguration configuration, String mode, IProgressMonitor monitor)
+		throws CoreException {
+		monitor.beginTask("", 1); //$NON-NLS-1$
+		try {
+			return Central.getWorkspace()
+				.readLocked(() -> {
+					IProject[] projects = getRelatedProjects();
+
+					monitor.subTask(DebugCoreMessages.LaunchConfigurationDelegate_6);
+					List<IAdaptable> errors = new ArrayList<>();
+					for (IProject project : projects) {
+						monitor.subTask(
+							MessageFormat.format(DebugCoreMessages.LaunchConfigurationDelegate_7, new Object[] {
+								project.getName()
+						}));
+						if (existsProblems(project)) {
+							errors.add(project);
+						}
+					}
```

Comment:
The preferred way to use a progress monitor is not to call `beginTask()`/`done()` directly, but to use a `SubMonitor`. From [here](https://wiki.eclipse.org/Progress_Reporting): 

> Whenever you write a method that accepts an IProgressMonitor, the first thing you should do is convert it to a SubMonitor using SubMonitor.convert(...)

We could also do a better job of reporting progress by setting the number of ticks equal to the number of `IProject`s and calling  `worked(1)` in the loop, also setting the status message to the name of the project. 99 times out of 100 this will probably not make much of a difference to the end user, but in that one time out of 100 it could be a useful debugging tool.